src/middlewared/middlewared/plugins/vm/vm_devices.py

```diff
@@ -55,14 +55,19 @@ async def disk_choices(self):
         """
         Returns disk choices for device type "DISK".
         """
-        return {
-            zvol_name_to_path(vol['id']): vol['id']
-            for vol in await self.middleware.call(
-                'pool.dataset.query', [['type', '=', 'VOLUME'], ['locked', '=', False]], {
-                    'extra': {'properties': ['encryption', 'keystatus']}
-                }
-            )
-        }
+        out = {}
+        zvols = await self.middleware.call(
+            'zfs.dataset.available_zvols_fast',
+            {"additional_information": ["ATTACHMENT"]}
+        )
+
+        for zvol in zvols.values():
+            if zvol['attachment']:
```

Comment:
It should be fine to use the same volume in multiple vms as long as it's not happening together, so i think we should not have this here.

Reply:
You mean it's safe as long as multiple VMs referencing it aren't running at the same time? Do we have validation to prevent this?

Reply:
Yup we should iirc, and if it's not happening - i can take care of that bit ;)

Reply:
Why would we allow configuration of the same zvol to 2 different VMs but disallow the starting of both VMs at the same time?

Reply:
So that the same disk can be used by both VMs ( not concurrently ) ?

Reply:
That's my point....why allow the config if the VMs can't be powered on concurrently? That seems like an edge-case. Even more so, the practicality of powering on a VM, writing data to the zvol, powering it down, then powering on the other one seems....strange.

Reply:
I don't understand ? So you are saying if i want to read data from a single disk on both VMs ( not concurrently ), i should powerdown the first one, remove the device from the first one and then attach the disk to the new VM and power it up and then repeat the process if i want to access vm 1 again ?

Reply:
I'm saying that allowing a zvol to be configured to 2 different VMs makes very little sense and I don't see any practical use-case for it.

Reply:
I don't see why we should explicitly disable that workflow also given it's already working ? So we are also changing behaviour in this case for no good reason to the best of my understanding ?

Reply:
Because it's setting the user up for data corruption. It doesn't matter if the workflow works right now or not. Giving the user a loaded gun without any safeguards is the incorrect way to design something.

Reply:
I don't follow, if the zvol is not concurrently being used by VMs, how is that causing data corruption ?

Reply:
How are you ensuring the data being written to the zvol is flushed to disk? How are you ensuring the zvol that is moved the other VM isn't formatted and overwritten instead of being mounted?

Reply:
For the second question, if a user is doing this he's managing the formatting part as he can't access it otherwise.

For the first question, how are your changes going to prevent the issue described ? User can still add the device to the second VM ( btw VMs are not running at this point, same applies to starting the other VM having same zvol ), how do we know the data has been flushed at that point to the disk ?

If you still think we should change the existing workflow, sure :)

Reply:
When a VM is powered down, the OS is responsible for flushing data to the disk before power off so it's "guaranteed" without any user intervention. So I'm coming full circle again to my original point. What is the practical use-case when VMs have to be powered down before making any changes?

1. Create VM1
2. Create Zvol1
3. Attach Zvol1 to VM1
4. VM1 makes changes to Zvol1
5. VM1 is powered down
6. Zvol1 is attached to VM2
7. User has to go into guest OS and mount the zvol inside the VM
8. Changes are made
9. VM2 is powered down
10. Zvol1 is changed to VM1
11. Etc etc

Is that really practical?

Reply:
So if on VM shutdown it's ensured data is flushed, we won't run into the case of data corruption as the second VM won't use the zvol unless the first one has been shutdown ? We should be good ?

Reply:
Yes, if the user can't change a zvol on a VM while it's powered on, then we don't have to worry about data being flushed.